File: benchmarks/bench_echo_server.cpp
// bench_echo_server.cpp — WebSocket echo server for remote network benchmarks

#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>
#include <wscpp/server.hpp>

using namespace wscpp;

namespace {

void usage(const char *argv0) {
    std::fprintf(stderr, "Usage: %s [port] [bind_host]\n", argv0);
    std::fprintf(stderr, "  Defaults: port=19081 bind=0.0.0.0\n");
}

} // namespace

int main(int argc, char *argv[]) {
    uint16_t port = 19081;
    std::string bind_host = "0.0.0.0";

    if (argc > 1) {
        if (std::string(argv[1]) == "-h" || std::string(argv[1]) == "--help") {
            usage(argv[0]);
            return 0;
        }
        port = static_cast<uint16_t>(std::atoi(argv[1]));
    }
    if (argc > 2) {
        bind_host = argv[2];
    }

    server srv;
    srv.set_on_connection([&](std::shared_ptr<connection> conn) {
        conn->set_on_message([conn](const std::vector<uint8_t> &data, frame::opcode op) {
            if (op == frame::opcode::BINARY) {
                conn->send_binary(data.data(), data.size());
            } else {
                conn->send_text(std::string(data.begin(), data.end()));
            }
        });
    });

    const std::error_code listen_ec = srv.listen(bind_host, port);
    if (listen_ec) {
        std::fprintf(stderr, "listen %s:%u failed: %s\n", bind_host.c_str(), port,
                     listen_ec.message().c_str());
        return 1;
    }
    const std::error_code start_ec = srv.start();
    if (start_ec) {
        std::fprintf(stderr, "start failed: %s\n", start_ec.message().c_str());
        return 1;
    }

    std::printf("wscpp echo server listening on %s:%u\n", bind_host.c_str(), port);
    srv.join();
    return 0;
}

File: rust/ws-rs-benches/src/bin/bench_echo_server.rs
// bench_echo_server — WebSocket echo server for remote network benchmarks

use std::env;
use std::sync::Arc;

use ws_rs::server::{echo_session, Server};

fn usage(argv0: &str) {
    eprintln!("Usage: {argv0} [port] [bind_host]");
    eprintln!("  Defaults: port=19081 bind=0.0.0.0");
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() > 1 && (args[1] == "-h" || args[1] == "--help") {
        usage(&args[0]);
        return;
    }

    let port: u16 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(19081);
    let bind_host = args.get(2).map(String::as_str).unwrap_or("0.0.0.0");

    let mut server = Server::new();
    if server.listen(bind_host, port).await.is_err() {
        eprintln!("listen {bind_host}:{port} failed");
        std::process::exit(1);
    }

    println!("ws-rs echo server listening on {bind_host}:{port}");

    loop {
        let accept = server.listener().unwrap().accept();
        let (stream, _) = match accept.await {
            Ok(v) => v,
            Err(_) => break,
        };
        let conn = match Server::complete_handshake(stream, None).await {
            Ok(c) => Arc::new(c),
            Err(_) => continue,
        };
        tokio::spawn(async move {
            let _ = echo_session(conn).await;
        });
    }
}
